src/Data/Variant/Internal.purs

```diff
@@ -1,25 +1,92 @@
 module Data.Variant.Internal
-  ( RLProxy(..)
-  , RProxy(..)
+  ( FProxy(..)
   , VariantCase
   , class VariantTags, variantTags
   , class Contractable, contractWith
+  , class VRMatching
+  , class RLMatch
+  , class VRFMatching
+  , class RLFMatch
   , lookupTag
   , lookupEq
   , lookupOrd
+  , module Exports
   ) where
 
 import Prelude
 import Control.Alternative (class Alternative, empty)
 import Data.List as L
-import Data.Symbol (SProxy(..), class IsSymbol, reflectSymbol)
+import Data.Symbol (class IsSymbol, SProxy(SProxy), reflectSymbol)
 import Data.Tuple (Tuple(..))
+import Data.Record.Unsafe (unsafeGet) as Exports
 import Partial.Unsafe (unsafeCrashWith)
 import Type.Row as R
+import Type.Row (RProxy, RLProxy(..))
+import Type.Row (RProxy(..), RLProxy(..)) as Exports
 
-data RProxy (r ∷ # Type) = RProxy
+-- | Proxy for a `Functor`.
+data FProxy (a ∷ Type → Type) = FProxy
 
-data RLProxy (rl ∷ R.RowList) = RLProxy
+-- | Type class that matches a row for a `record` that will eliminate a row for
+-- | a `variant`, producing a `result` of the specified type. Just a wrapper for
+-- | `RLMatch` to convert `RowToList` and vice versa.
+class VRMatching
+    (variant ∷ # Type)
+    (record ∷ # Type)
+    result
+  | → variant record result
+instance variantRecordMatching
+  ∷ ( R.RowToList variant vlist
+    , R.RowToList record rlist
+    , RLMatch vlist rlist result
+    , R.ListToRow vlist variant
+    , R.ListToRow rlist record )
+  ⇒ VRMatching variant record result
+
+-- | Checks that a `RowList` matches the argument to be given to the function
+-- | in the other `RowList` with the same label, such that it will produce the
+-- | result type.
+class RLMatch
```

Comment:
Okay, good point! It's a pity that this is overlapping though:
```purescript
instance variantMatchAFail
  ∷ Fail (TypeConcat "Unhandled variant case " (TypeConcat sym (TypeConcat " of type " (TypeString a))))
  ⇒ VariantMatchEachCase (R.Cons sym a v) R.Nil res
```

```
[1/1 OverlappingInstances] 
  Overlapping type class instances found for
    Data.Variant.Internal.VariantMatchEachCase (Cons "bar" String (Cons "baz" Boolean (Cons "foo" Int Nil)))
                                               Nil
                                               String

  The following instances were found:
    Data.Variant.Internal.variantMatchAFail (chosen)
    Data.Variant.Internal.variantMatchCons
    Data.Variant.Internal.variantMatchNil

[1/1 NoInstanceFound] test/Variant.purs:62:14
                   v
  62      match' = match
  63        { }
              ^
  A custom type error occurred while solving type class constraints:
    Unhandled variant case bar of type String
```

Looks nice, but it only seemed to work when the record is empty ... Hopefully the errors with `Cons` are sufficiently obvious.